File: src/main/java/io/github/shirohoo/library/domain/user/UserRepository.java
package io.github.shirohoo.library.domain.user;

import io.github.shirohoo.library.application.domain.user.User;
import java.util.List;
import java.util.Optional;

public interface UserRepository {
    User save(User newUser);

    List<User> findAll();

    Optional<User> findBy(long id);

    Optional<User> findBy(String username);

    boolean delete(User user);
}

File: src/main/kotlin/io/github/shirohoo/library/domain/user/UserRepository.kt
package io.github.shirohoo.library.domain.user

interface UserRepository {
    fun save(newUser: User): User
    fun findAll(): List<User>
    fun findAllWithHistories(): List<User>
    fun findBy(id: Long): User
    fun findBy(username: String): User
    fun delete(user: User)
}